src/main/java/seedu/address/logic/commands/EditCommand.java

```diff
@@ -72,22 +72,51 @@ public CommandResult execute(Model model) throws CommandException {
         requireNonNull(model);
         List<Person> lastShownList = model.getFilteredPersonList();
 
-        if (index.getZeroBased() >= lastShownList.size()) {
-            throw new CommandException(Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
-        }
+        // Find the person to edit based on name or ID
+        Optional<Person> personOptional = findPersonToEdit(lastShownList);
 
-        Person personToEdit = lastShownList.get(index.getZeroBased());
+        Person personToEdit = personOptional.get();
         Person editedPerson = createEditedPerson(personToEdit, editPersonDescriptor);
 
-        if (!personToEdit.isSamePerson(editedPerson) && model.hasPerson(editedPerson)) {
-            throw new CommandException(MESSAGE_DUPLICATE_PERSON);
-        }
-
         model.setPerson(personToEdit, editedPerson);
         model.updateFilteredPersonList(PREDICATE_SHOW_ALL_PERSONS);
         return new CommandResult(String.format(MESSAGE_EDIT_PERSON_SUCCESS, Messages.format(editedPerson)));
     }
 
+    /**
+     * Finds the person to edit based on the provided name and/or NRIC.
+     *
+     * @param persons A list of persons to search within.
+     * @return An Optional containing the person to edit, if found, or an empty Optional if not found.
+     * @throws CommandException if there is inconsistency between the provided name and NRIC.
+     */
+    public Optional<Person> findPersonToEdit(List<Person> persons) throws CommandException {
+        if (name != null && nric != null) {
+            // Search for a person by name and NRIC
+            Optional<Person> personByName = persons.stream()
+                    .filter(person -> person.getName().equals(name))
+                    .findFirst();
+            Optional<Person> personByNric = persons.stream()
+                    .filter(person -> person.getNric().equals(nric))
+                    .findFirst();
+            // Check if both Optional instances are not empty, and return the one that represents the same person
+            if (personByName.isPresent() && personByNric.isPresent() && personByName.get() == personByNric.get()) {
+                return personByName;
+            } else {
+                throw new CommandException(MESSAGE_INCONSISTENT_NAME_AND_ID);
+            }
+        }
+        if (name != null) {
+            return persons.stream()
+                    .filter(person -> person.getName().equals(name))
+                    .findFirst();
+        } else {
+            return persons.stream()
+                    .filter(person -> person.getNric().equals(nric))
+                    .findFirst();
+        }
+    }
+
```

Comment:
This makes sense currently to keep isolated in `EditCommand` since we do not have a method to find a user by name/NRIC right now, though this is more suited for `FindCommand`. We will need to flag this method for refactoring at some point in the future.